From 5dd2e5a4162d0a53655317e50d37f162645ae004 Mon Sep 17 00:00:00 2001
From: Harman Kalra <hkalra@marvell.com>
Date: Fri, 1 Dec 2023 01:32:15 +0530
Subject: [PATCH 800/955] common/cnxk: base NPC changes for representor ports

Adding base NPC support for represented ports and introducing a
predefined index for NIXX_AF_LFX_RX_VTAG_TYPEX, which can perform
strip action in case of flow installed by PF on behalf of a
representor port

Change-Id: If15330b44d26e37926e3524d7063fa124e166ece
Signed-off-by: Harman Kalra <hkalra@marvell.com>
---
 drivers/common/cnxk/roc_npc.c       |  2 ++
 drivers/common/cnxk/roc_npc_parse.c | 17 +++++++++++++++--
 drivers/common/cnxk/roc_npc_priv.h  |  1 +
 3 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/common/cnxk/roc_npc.c b/drivers/common/cnxk/roc_npc.c
index 45056e255c64c..2f85f0c683aa6 100644
--- a/drivers/common/cnxk/roc_npc.c
+++ b/drivers/common/cnxk/roc_npc.c
@@ -1390,6 +1390,7 @@ npc_vtag_strip_action_configure(struct mbox *mbox,
 
 	rx_vtag_action |= (NIX_RX_VTAGACTION_VTAG_VALID << 15);
 	rx_vtag_action |= ((uint64_t)NPC_LID_LB << 8);
+	rx_vtag_action |= ((uint64_t)NIX_RX_VTAG_TYPE6 << 12);
 	rx_vtag_action |= NIX_RX_VTAGACTION_VTAG0_RELPTR;
 
 	if (*strip_cnt == 2) {
@@ -1440,6 +1441,7 @@ npc_vtag_action_program(struct roc_npc *roc_npc,
 			if (rc)
 				return rc;
 
+			plt_npc_dbg("VLAN strip action, strip_cnt %d", strip_cnt);
 			if (strip_cnt == 2)
 				actions++;
 
diff --git a/drivers/common/cnxk/roc_npc_parse.c b/drivers/common/cnxk/roc_npc_parse.c
index aa20a3ab74faf..81cbb816373a2 100644
--- a/drivers/common/cnxk/roc_npc_parse.c
+++ b/drivers/common/cnxk/roc_npc_parse.c
@@ -46,11 +46,24 @@ npc_parse_port_representor_id(struct npc_parse_state *pst)
 	return 0;
 }
 
+int
+npc_parse_represented_port_id(struct npc_parse_state *pst)
+{
+	if (pst->pattern->type != ROC_NPC_ITEM_TYPE_REPRESENTED_PORT)
+		return 0;
+
+	if (pst->flow->nix_intf != NIX_INTF_RX)
+		return -EINVAL;
+
+	pst->pattern++;
+
+	return 0;
+}
+
 static int
 npc_flow_raw_item_prepare(const struct roc_npc_flow_item_raw *raw_spec,
 			  const struct roc_npc_flow_item_raw *raw_mask,
-			  struct npc_parse_item_info *info, uint8_t *spec_buf,
-			  uint8_t *mask_buf)
+			  struct npc_parse_item_info *info, uint8_t *spec_buf, uint8_t *mask_buf)
 {
 
 	memset(spec_buf, 0, NPC_MAX_RAW_ITEM_LEN);
diff --git a/drivers/common/cnxk/roc_npc_priv.h b/drivers/common/cnxk/roc_npc_priv.h
index f29a612c2ecff..77bbbecd22ddd 100644
--- a/drivers/common/cnxk/roc_npc_priv.h
+++ b/drivers/common/cnxk/roc_npc_priv.h
@@ -458,6 +458,7 @@ int npc_parse_item_basic(const struct roc_npc_item_info *item, struct npc_parse_
 int npc_parse_meta_items(struct npc_parse_state *pst);
 int npc_parse_mark_item(struct npc_parse_state *pst);
 int npc_parse_port_representor_id(struct npc_parse_state *pst);
+int npc_parse_represented_port_id(struct npc_parse_state *pst);
 int npc_parse_pre_l2(struct npc_parse_state *pst);
 int npc_parse_higig2_hdr(struct npc_parse_state *pst);
 int npc_parse_cpt_hdr(struct npc_parse_state *pst);
-- 
2.25.1

